### ISSUE TITLE
Deepcopy of model with categorical features causes infinite hang during prediction after refitting

### ISSUE THREAD

**Issue by MihailSeregin** [catboost/catboost#2905]
**Problem**
Модель виснет на этапе инференса. 

**Описание**
В  коде копируем исходную модель, переобучаем  и делаем предсказание, на котором часто зависает.  
Без категориальных признаков отрабатывает всегда. Ниже привожу простой пример,  на котором зависает выполнение. Есть подозрение на строчку  deepcopy(model), и  .drop(features_to_drop,  axis=1). 

```python
import pandas as pd
import numpy as np
from catboost import CatBoostClassifier, Pool
from sklearn.model_selection import train_test_split
from copy import deepcopy

np.random.seed(0)

n_samples = 40*10**3

num_features = pd.DataFrame({
    'num_1': np.random.randn(n_samples),
    'num_2': np.random.rand(n_samples) * 100,
    'num_3': np.random.randint(0, 50, n_samples)
})


cat_features = pd.DataFrame({
    'cat_1': np.random.choice(['A', 'B', 'C'], n_samples),
    'cat_2': np.random.choice(['X', 'Y'], n_samples),
    'cat_3': np.random.choice(['red', 'green', 'blue', 'yellow'], n_samples)
})


target = ((num_features['num_1'] + 
           (cat_features['cat_1'] == 'B').astype(int) - 
           (cat_features['cat_3'] == 'green').astype(int) + 
           np.random.randn(n_samples)*0.5) > 0).astype(int)

df = pd.concat([num_features, cat_features], axis=1)

df_train, df_test, target_train, target_test = train_test_split(df, target, test_size=0.3, random_state=42)

cat_features_indices = [df_train.columns.get_loc(col) for col in ['cat_1', 'cat_2', 'cat_3']]

model = CatBoostClassifier(
    iterations=40,
    learning_rate=0.25,
    depth=11,
    verbose=False,
    random_seed=42, 
    class_weights=[1, 20],
    thread_count=3, 
    l2_leaf_reg=50
)

model.fit(
    df_train, target_train,
    cat_features=cat_features_indices
)

features_to_drop = []
def refit_and_predict(model, df_train, target_train, df_test, features_to_drop, cat_features_indices):
    from copy import deepcopy

    model_copy = deepcopy(model)

    model_copy = model_copy.fit(
        df_train.drop(features_to_drop, axis=1),
        target_train,
        cat_features=cat_features_indices   
    )
    

    # **problem here**
    refit_preds_train = model_copy.predict(df_train.drop(features_to_drop, axis=1))
    refit_preds_test = model_copy.predict(df_test.drop(features_to_drop, axis=1))

    return refit_preds_train, refit_preds_test
    

train_preds, test_preds = refit_and_predict(model, df_train, target_train, df_test, features_to_drop, cat_features_indices)
```
###requirements
pandas==1.4.3
numpy==1.26.4
catboost==1.2.8
scikit-learn==1.5.2
ipykernel==6.29.5

version of python  3.9
os: SberLinux

**Comment by aghabidareh** [comment]
instead of deepcopying the trained model, you should create a new CatBoostClassifier with the same parameters and fit it again, for example: 
```python
def refit_and_predict(model, df_train, target_train, df_test, features_to_drop, cat_features_indices):
    # Recreate model with same params instead of deepcopy
    model_copy = CatBoostClassifier(**model.get_params())
    
    model_copy.fit(
        df_train.drop(features_to_drop, axis=1),
        target_train,
        cat_features=cat_features_indices,
        verbose=False
    )
    
    refit_preds_train = model_copy.predict(df_train.drop(features_to_drop, axis=1))
    refit_preds_test = model_copy.predict(df_test.drop(features_to_drop, axis=1))
    return refit_preds_train, refit_preds_test

```

also you can save the trained model and again load it, if you really need it
example:
```python
model.save_model("temp.cbm")
model_copy = CatBoostClassifier().load_model("temp.cbm")
```
